src/contracts/atlas/Escrow.sol

```diff
@@ -430,57 +422,147 @@ abstract contract Escrow is AtlETH {
-            return 1 << uint256(SolverOutcome.PostSolverFailed);
-        } else if (errorSwitch == IntentUnfulfilled.selector) {
-            return 1 << uint256(SolverOutcome.IntentUnfulfilled);
-        } else if (errorSwitch == SolverBidUnpaid.selector) {
-            return 1 << uint256(SolverOutcome.BidNotPaid);
-        } else if (errorSwitch == BalanceNotReconciled.selector) {
-            return 1 << uint256(SolverOutcome.BalanceNotReconciled);
-        } else if (errorSwitch == InvalidEntry.selector) {
-            // DAppControl is attacking solver contract - treat as AlteredControl
-            return 1 << uint256(SolverOutcome.AlteredControl);
-        } else if (errorSwitch == CallbackNotCalled.selector) {
-            return 1 << uint256(SolverOutcome.SolverOpReverted);
+
+        // Update solverTracker with returned data
+        solverTracker = abi.decode(data, (SolverTracker));
+
+        // ------------------------------------- //
+        //              Solver Call              //
+        // ------------------------------------- //
+
+        // Set the solver lock - will perform accounting checks if value borrowed. If `_trySolverLock()` returns false,
+        // we revert here and catch the error in `_solverOpWrapper()` above
+        if (!_trySolverLock(solverOp)) revert InsufficientEscrow();
+
+        data = abi.encodeCall(
+            ISolverContract.atlasSolverCall,
+            (solverOp.from, key.executionEnvironment, solverOp.bidToken, bidAmount, solverOp.data, dAppReturnData)
+        );
+        (success,) = solverOp.solver.call{ value: solverOp.value, gas: gasLimit }(data);
+
+        if (!success) revert SolverOpReverted();
+
+        // ------------------------------------- //
+        //            Post-Solver Call           //
+        // ------------------------------------- //
+
+        data = abi.encodeCall(IExecutionEnvironment.solverPostTryCatch, (solverOp, dAppReturnData, solverTracker));
+        (success, data) = key.executionEnvironment.call(abi.encodePacked(data, key.holdPostSolverLock().pack()));
+
+        // If ExecutionEnvironment.solverPostTryCatch() failed, bubble up the error
+        if (!success) {
+            assembly {
+                revert(add(data, 32), mload(data))
+            }
         }
-        return 1 << uint256(SolverOutcome.EVMError);
+
+        // Update solverTracker with returned data
+        solverTracker = abi.decode(data, (SolverTracker));
+
+        // Verify that the solver repaid their borrowed solverOp.value by calling `reconcile()`. If solver did not fully
+        // repay via `reconcile()`, the postSolverCall may still have covered the outstanding debt via `contribute()` so
+        // we do a final repayment check here.
+        (, calledback, success) = _solverLockData();
+
+        if (!calledback) revert CallbackNotCalled();
+        if (!success && deposits < claims + withdrawals) revert BalanceNotReconciled();
```

Comment:
Yes it still uses the fulfilled flag - but the idea is that it should be reliable now. As of the Lock PR, borrow is limited to SolverOperations phase or before, and `reconcile()` can only be called during the SolverOperations phase. So, if a solver calls `reconcile()` and sets `fulfilled = true` then we know it can't go back to unfulfilled as the dapp can no longer borrow - only repay further via `contribute()`